### ISSUE TITLE
Enabling cross cluster communication in k8s

### ISSUE THREAD

**Issue by neith00** [weaveworks/weave#2973]
When using weave with k8s to connect two k8s clusters, the default iptables configuration prevent traffic going from pods on cluster one to pods of cluster two.
This PR https://github.com/weaveworks/weave/pull/2972 fix the iptables rules to fix this issue.

**Comment by bboreham** [comment]
A connection should be allowed if the namespace is in `DefaultAllow` mode (the default), or by a more specific [network policy](https://kubernetes.io/docs/concepts/services-networking/networkpolicies/).

Since you mention two clusters, you are perhaps hitting the limitation that Kubernetes Network Policies can only refer specifically to one cluster.
You can write non-specific rules, opening all connections to a port.

Or we could consider some extension to Kubernetes Network Policies to support your use-case.

Please can you explain more about what you want to do.

**Comment by neith00** [comment]
As you ask on slack here is the iptables output
```
# Generated by iptables-save v1.6.0 on Wed May 17 11:07:04 2017
*mangle
:PREROUTING ACCEPT [7553478:1939413210]
:INPUT ACCEPT [7553474:1939412922]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [7587220:1982189840]
:POSTROUTING ACCEPT [7587217:1982189660]
:WEAVE-IPSEC-IN - [0:0]
:WEAVE-IPSEC-IN-MARK - [0:0]
:WEAVE-IPSEC-OUT - [0:0]
:WEAVE-IPSEC-OUT-MARK - [0:0]
-A INPUT -j WEAVE-IPSEC-IN
-A OUTPUT -j WEAVE-IPSEC-OUT
-A WEAVE-IPSEC-IN -s 192.168.204.30/32 -d 192.168.204.27/32 -p esp -m esp --espspi 3646789848 -j WEAVE-IPSEC-IN-MARK
-A WEAVE-IPSEC-IN-MARK -j MARK --set-xmark 0x20000/0x20000
-A WEAVE-IPSEC-OUT -s 192.168.204.27/32 -d 192.168.204.30/32 -p udp -m udp --dport 6784 -j WEAVE-IPSEC-OUT-MARK
-A WEAVE-IPSEC-OUT-MARK -j MARK --set-xmark 0x20000/0x20000
COMMIT
# Completed on Wed May 17 11:07:04 2017
# Generated by iptables-save v1.6.0 on Wed May 17 11:07:04 2017
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:DOCKER - [0:0]
:KUBE-MARK-DROP - [0:0]
:KUBE-MARK-MASQ - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-SEP-ERGYM4I7URZVCEUS - [0:0]
:KUBE-SEP-N2H2RTMJWQXTOA35 - [0:0]
:KUBE-SEP-NJX6LAMUYAUENME6 - [0:0]
:KUBE-SEP-V43MXM5NC7UWXX2F - [0:0]
:KUBE-SEP-ZV2PEOV4IHFALTUC - [0:0]
:KUBE-SERVICES - [0:0]
:KUBE-SVC-BDJ6JTTI7YL7IHTD - [0:0]
:KUBE-SVC-CIGGPN5BGHD4S2O2 - [0:0]
:KUBE-SVC-NPX46M4PTMTKRN6Y - [0:0]
:KUBE-SVC-U7MHD6KS7NJ6FFYZ - [0:0]
:KUBE-SVC-UC7ZWITLXDTOOKDD - [0:0]
:WEAVE - [0:0]
-A PREROUTING -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -m comment --comment "kubernetes postrouting rules" -j KUBE-POSTROUTING
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
-A POSTROUTING -j WEAVE
-A DOCKER -i docker0 -j RETURN
-A KUBE-MARK-DROP -j MARK --set-xmark 0x8000/0x8000
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x4000/0x4000
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -m mark --mark 0x4000/0x4000 -j MASQUERADE
-A KUBE-SEP-ERGYM4I7URZVCEUS -s 192.168.204.27/32 -m comment --comment "default/kubernetes:https" -j KUBE-MARK-MASQ
-A KUBE-SEP-ERGYM4I7URZVCEUS -p tcp -m comment --comment "default/kubernetes:https" -m recent --set --name KUBE-SEP-ERGYM4I7URZVCEUS --mask 255.255.255.255 --rsource -m tcp -j DNAT --to-destination 192.168.204.27:6443
-A KUBE-SEP-N2H2RTMJWQXTOA35 -s 10.233.96.3/32 -m comment --comment "kube-system/kubedns:dns-tcp" -j KUBE-MARK-MASQ
-A KUBE-SEP-N2H2RTMJWQXTOA35 -p tcp -m comment --comment "kube-system/kubedns:dns-tcp" -m tcp -j DNAT --to-destination 10.233.96.3:53
-A KUBE-SEP-NJX6LAMUYAUENME6 -s 10.233.96.0/32 -m comment --comment "kube-system/dnsmasq:dns" -j KUBE-MARK-MASQ
-A KUBE-SEP-NJX6LAMUYAUENME6 -p udp -m comment --comment "kube-system/dnsmasq:dns" -m udp -j DNAT --to-destination 10.233.96.0:53
-A KUBE-SEP-V43MXM5NC7UWXX2F -s 10.233.96.0/32 -m comment --comment "kube-system/dnsmasq:dns-tcp" -j KUBE-MARK-MASQ
-A KUBE-SEP-V43MXM5NC7UWXX2F -p tcp -m comment --comment "kube-system/dnsmasq:dns-tcp" -m tcp -j DNAT --to-destination 10.233.96.0:53
-A KUBE-SEP-ZV2PEOV4IHFALTUC -s 10.233.96.3/32 -m comment --comment "kube-system/kubedns:dns" -j KUBE-MARK-MASQ
-A KUBE-SEP-ZV2PEOV4IHFALTUC -p udp -m comment --comment "kube-system/kubedns:dns" -m udp -j DNAT --to-destination 10.233.96.3:53
-A KUBE-SERVICES ! -s 10.233.64.0/18 -d 10.233.0.2/32 -p tcp -m comment --comment "kube-system/dnsmasq:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.233.0.2/32 -p tcp -m comment --comment "kube-system/dnsmasq:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-CIGGPN5BGHD4S2O2
-A KUBE-SERVICES ! -s 10.233.64.0/18 -d 10.233.0.2/32 -p udp -m comment --comment "kube-system/dnsmasq:dns cluster IP" -m udp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.233.0.2/32 -p udp -m comment --comment "kube-system/dnsmasq:dns cluster IP" -m udp --dport 53 -j KUBE-SVC-UC7ZWITLXDTOOKDD
-A KUBE-SERVICES ! -s 10.233.64.0/18 -d 10.233.0.3/32 -p udp -m comment --comment "kube-system/kubedns:dns cluster IP" -m udp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.233.0.3/32 -p udp -m comment --comment "kube-system/kubedns:dns cluster IP" -m udp --dport 53 -j KUBE-SVC-U7MHD6KS7NJ6FFYZ
-A KUBE-SERVICES ! -s 10.233.64.0/18 -d 10.233.0.3/32 -p tcp -m comment --comment "kube-system/kubedns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.233.0.3/32 -p tcp -m comment --comment "kube-system/kubedns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-BDJ6JTTI7YL7IHTD
-A KUBE-SERVICES ! -s 10.233.64.0/18 -d 10.233.0.1/32 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d 10.233.0.1/32 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-SVC-NPX46M4PTMTKRN6Y
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
-A KUBE-SVC-BDJ6JTTI7YL7IHTD -m comment --comment "kube-system/kubedns:dns-tcp" -j KUBE-SEP-N2H2RTMJWQXTOA35
-A KUBE-SVC-CIGGPN5BGHD4S2O2 -m comment --comment "kube-system/dnsmasq:dns-tcp" -j KUBE-SEP-V43MXM5NC7UWXX2F
-A KUBE-SVC-NPX46M4PTMTKRN6Y -m comment --comment "default/kubernetes:https" -m recent --rcheck --seconds 10800 --reap --name KUBE-SEP-ERGYM4I7URZVCEUS --mask 255.255.255.255 --rsource -j KUBE-SEP-ERGYM4I7URZVCEUS
-A KUBE-SVC-NPX46M4PTMTKRN6Y -m comment --comment "default/kubernetes:https" -j KUBE-SEP-ERGYM4I7URZVCEUS
-A KUBE-SVC-U7MHD6KS7NJ6FFYZ -m comment --comment "kube-system/kubedns:dns" -j KUBE-SEP-ZV2PEOV4IHFALTUC
-A KUBE-SVC-UC7ZWITLXDTOOKDD -m comment --comment "kube-system/dnsmasq:dns" -j KUBE-SEP-NJX6LAMUYAUENME6
-A WEAVE -s 10.233.64.0/18 -d 224.0.0.0/4 -j RETURN
-A WEAVE ! -s 10.233.64.0/18 -d 10.233.64.0/18 -j MASQUERADE
-A WEAVE -s 10.233.64.0/18 ! -d 10.233.64.0/18 -j MASQUERADE
COMMIT
# Completed on Wed May 17 11:07:04 2017
# Generated by iptables-save v1.6.0 on Wed May 17 11:07:04 2017
*filter
:INPUT ACCEPT [6:437]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [5:387]
:DOCKER - [0:0]
:DOCKER-ISOLATION - [0:0]
:KUBE-FIREWALL - [0:0]
:KUBE-SERVICES - [0:0]
:WEAVE-IPSEC-IN - [0:0]
:WEAVE-NPC - [0:0]
:WEAVE-NPC-DEFAULT - [0:0]
:WEAVE-NPC-INGRESS - [0:0]
-A INPUT -j KUBE-FIREWALL
-A INPUT -d 172.17.0.1/32 -i docker0 -p tcp -m tcp --dport 6783 -j DROP
-A INPUT -d 172.17.0.1/32 -i docker0 -p udp -m udp --dport 6783 -j DROP
-A INPUT -d 172.17.0.1/32 -i docker0 -p udp -m udp --dport 6784 -j DROP
-A INPUT -i docker0 -p udp -m udp --dport 53 -j ACCEPT
-A INPUT -i docker0 -p tcp -m tcp --dport 53 -j ACCEPT
-A INPUT -j WEAVE-IPSEC-IN
-A FORWARD -i docker0 -o weave -j DROP
-A FORWARD -j DOCKER-ISOLATION
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A FORWARD -o weave -j WEAVE-NPC
-A FORWARD -o weave -m state --state NEW -j NFLOG --nflog-group 86
-A FORWARD -o weave -j DROP
-A FORWARD -i weave ! -o weave -j ACCEPT
-A FORWARD -o weave -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT -j KUBE-FIREWALL
-A OUTPUT ! -p esp -m policy --dir out --pol none -m mark --mark 0x20000/0x20000 -j DROP
-A DOCKER-ISOLATION -j RETURN
-A KUBE-FIREWALL -m comment --comment "kubernetes firewall for dropping marked packets" -m mark --mark 0x8000/0x8000 -j DROP
-A WEAVE-IPSEC-IN -s 192.168.204.30/32 -d 192.168.204.27/32 -p udp -m udp --dport 6784 -m mark ! --mark 0x20000/0x20000 -j DROP
-A WEAVE-NPC -m state --state RELATED,ESTABLISHED -j ACCEPT
-A WEAVE-NPC -d 224.0.0.0/4 -j ACCEPT
-A WEAVE-NPC -m state --state NEW -j WEAVE-NPC-DEFAULT
-A WEAVE-NPC -m state --state NEW -j WEAVE-NPC-INGRESS
-A WEAVE-NPC-DEFAULT -m set --match-set weave-k?Z;25^M}|1s7P3|H9i;*;MhG dst -j ACCEPT
-A WEAVE-NPC-DEFAULT -m set --match-set weave-iuZcey(5DeXbzgRFs8Szo]<@p dst -j ACCEPT
COMMIT
# Completed on Wed May 17 11:07:04 2017
```

**Comment by bboreham** [comment]
ok, so a packet for a new connection should follow these rules:

```
-A FORWARD -o weave -j WEAVE-NPC
-A WEAVE-NPC -m state --state NEW -j WEAVE-NPC-DEFAULT
-A WEAVE-NPC-DEFAULT -m set --match-set weave-k?Z;25^M}|1s7P3|H9i;*;MhG dst -j ACCEPT
-A WEAVE-NPC-DEFAULT -m set --match-set weave-iuZcey(5DeXbzgRFs8Szo]<@p dst -j ACCEPT
```

and if the destination pod's address is in either of those ipsets it will be accepted.
If you have no networking policies in place then it should be in a set.

You can check the contents by doing:

    ipset list

or, if you don't have that installed, run the one inside the Weave-NPC container:

    docker run --rm --privileged --net=host --entrypoint=/usr/sbin/ipset weaveworks/weave-npc:1.9.5 list

You can also check the logs of the `weave-npc` container on that host; careful they are very verbose.

**Comment by neith00** [comment]
infos from first cluster:
```
ubuntu@ubuntu:~$ kubectl get pods -o wide --all-namespaces 
NAMESPACE     NAME                                  READY     STATUS    RESTARTS   AGE       IP               NODE
default       busybox                               1/1       Running   0          34m       10.233.96.5      ubuntu1
kube-system   dnsmasq-411420702-pl72k               1/1       Running   0          50m       10.233.96.1      ubuntu1
kube-system   dnsmasq-autoscaler-1155841093-1zhrm   1/1       Running   0          50m       10.233.96.2      ubuntu1
kube-system   kube-apiserver-ubuntu                 1/1       Running   0          51m       192.168.204.34   ubuntu
kube-system   kube-controller-manager-ubuntu        1/1       Running   0          50m       192.168.204.34   ubuntu
kube-system   kube-proxy-ubuntu                     1/1       Running   0          50m       192.168.204.34   ubuntu
kube-system   kube-proxy-ubuntu1                    1/1       Running   0          49m       192.168.204.17   ubuntu1
kube-system   kube-scheduler-ubuntu                 1/1       Running   0          50m       192.168.204.34   ubuntu
kube-system   kubedns-3830354952-r8vw6              3/3       Running   0          49m       10.233.96.3      ubuntu1
kube-system   kubedns-autoscaler-54374881-l6w5l     1/1       Running   0          49m       10.233.96.4      ubuntu1
kube-system   nginx-proxy-ubuntu1                   1/1       Running   0          49m       192.168.204.17   ubuntu1
kube-system   weave-net-879t1                       2/2       Running   0          50m       192.168.204.17   ubuntu1
kube-system   weave-net-wfdhg                       2/2       Running   0          50m       192.168.204.34   ubuntu
ubuntu@ubuntu:~$ kubectl --namespace=kube-system exec -t weave-net-879t1 /usr/sbin/ipset list
Defaulting container name to weave.
Use 'kubectl describe pod/weave-net-879t1' to see all of the containers in this pod.
Name: weave-k?Z;25^M}|1s7P3|H9i;*;MhG
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 176
References: 1
Members:
10.233.96.5
 
Name: weave-iuZcey(5DeXbzgRFs8Szo]<@p
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 368
References: 1
Members:
10.233.96.4
10.233.96.3
10.233.96.2
10.233.96.1
10.233.104.5
```

Infos from second cluster:

```
root@ubuntu:~# kubectl get pods --all-namespaces -o wide
NAMESPACE     NAME                                  READY     STATUS    RESTARTS   AGE       IP               NODE
default       busybox                               1/1       Running   0          35m       10.233.104.5     ubuntu1
kube-system   dnsmasq-411420702-27lfr               1/1       Running   0          37m       10.233.104.2     ubuntu1
kube-system   dnsmasq-autoscaler-1155841093-xfgq7   1/1       Running   0          37m       10.233.104.1     ubuntu1
kube-system   kube-apiserver-ubuntu                 1/1       Running   0          38m       192.168.204.15   ubuntu
kube-system   kube-controller-manager-ubuntu        1/1       Running   0          37m       192.168.204.15   ubuntu
kube-system   kube-proxy-ubuntu                     1/1       Running   0          37m       192.168.204.15   ubuntu
kube-system   kube-proxy-ubuntu1                    1/1       Running   0          37m       192.168.204.36   ubuntu1
kube-system   kube-scheduler-ubuntu                 1/1       Running   0          37m       192.168.204.15   ubuntu
kube-system   kubedns-3830354952-ldjhj              3/3       Running   0          36m       10.233.104.3     ubuntu1
kube-system   kubedns-autoscaler-54374881-fx55d     1/1       Running   0          36m       10.233.104.4     ubuntu1
kube-system   nginx-proxy-ubuntu1                   1/1       Running   0          37m       192.168.204.36   ubuntu1
kube-system   weave-net-p3v84                       2/2       Running   0          37m       192.168.204.36   ubuntu1
kube-system   weave-net-zpwj9                       2/2       Running   0          37m       192.168.204.15   ubuntu
root@ubuntu:~# kubectl --namespace=kube-system exec -t weave-net-p3v84 /usr/sbin/ipset list
Defaulting container name to weave.
Use 'kubectl describe pod/weave-net-p3v84' to see all of the containers in this pod.
Name: weave-k?Z;25^M}|1s7P3|H9i;*;MhG
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 176
References: 1
Members:
10.233.104.5

Name: weave-iuZcey(5DeXbzgRFs8Szo]<@p
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 368
References: 1
Members:
10.233.96.5
10.233.104.2
10.233.104.3
10.233.104.1
10.233.104.4
```

**Comment by bboreham** [comment]
Based on further conversation on Slack and various experiments, here is what I think is going wrong:

- Since there are no policies in effect, weave-npc adds all pod addresses to the "allowed destinations" ipset.
- Since weave-npc runs separately in each Kubernetes cluster, each one only adds addresses from the local cluster.
- However, the rules are applied at both ends of a connection, so they block packets from leaving one cluster destined to the other cluster.

A fix is being developed at #2979.  In the meantime you could disable weave-npc - see https://www.weave.works/docs/net/latest/kube-addon/#configuration-options

**Comment by marccarre** [comment]
See also discussion starting [here](https://weave-community.slack.com/archives/C2ND76PAA/p1495184414533203) on Weave Community's Slack.

**Comment by neith00** [comment]
Thanks, we are currently disabling npc and looking forward for the new release.

**Comment by bboreham** [comment]
Fix is now out in [v1.9.6](https://github.com/weaveworks/weave/releases/v1.9.6)
